src/io.rs

```diff
@@ -2,85 +2,126 @@
 
 use core::{fmt, slice};
 use core::fmt::Write;
+use core::ptr;
 
 /// File descriptors
-const STDOUT: usize = 1;
-const STDERR: usize = 2;
+static mut STDOUT: isize = -1;
+static mut STDERR: isize = -1;
 
 /// Host's standard error
 struct Stderr;
 
 /// Host's standard output
 struct Stdout;
 
-fn write_all(fd: usize, mut buffer: &[u8]) {
+pub fn get_stdout() -> isize {
+    // Safe: 32-bit accesses are atomic on ARM
+    unsafe{ptr::read_volatile(&STDOUT)}
+}
+
+pub fn get_stderr() -> isize {
+    // Safe: 32-bit accesses are atomic on ARM
+    unsafe{ptr::read_volatile(&STDERR)}
+}
+
+/// Open stdout and stderr.
+pub fn open_streams() -> Result<(),()> {
```

Comment:
If we go this route then I don't know if it's worth to cache the file descriptors of stdout and stderr. I assume it's OK to open e.g. stdout more than once since this function doesn't check if stdout has already been opened.

Caching the file descriptors does save us a semihosting call which I assume is slow but then all semihosting calls are slow so maybe it doesn't matter that much as you won't be use semihosting anywhere where timing is critical anyway.

Not caching saves us the use of `static` variables and volatile operations.